### PR TITLE
iProc pax dma driver support

### PR DESCRIPTION
**Pull request by raveenp**
This patch set series adds support for Broadcom PAX(PCIE<->AXI) DMA driver.

### CODE REVIEW

File: drivers/dma/CMakeLists.txt
Comment (olofj) [inline drivers/dma/CMakeLists.txt:16]: Hi, I think you should reshuffle the config options such that you don't have a repeated linke with the C file twice here (see below).

File: drivers/dma/Kconfig.iproc_pax
Comment (olofj) [inline drivers/dma/Kconfig.iproc_pax:14]: If you make DMA_IPROC_PAX_V2 a sub-option, and make it depend on DMA_IPROC_PAX, you will only need one Cmake entry, and you can make the options below depend on DMA_IPROC_PAX instead of doing an if/endif section.

File: drivers/dma/dma_iproc_pax_common.h
Comment (olofj) [inline drivers/dma/dma_iproc_pax_common.h:15]: Nit: Lining up the values in one column tends to be slightly easier to read, but not a huge deal either way.

Comment (olofj) [reply drivers/dma/dma_iproc_pax_common.h:15]: Also, overall, some of the names here are pretty generic and not namespaced with a prefix (PAX_* or somesuch). It might be useful to add that to avoid conflicts at some point.

File: drivers/dma/dma_iproc_pax_common.h
Comment (olofj) [inline drivers/dma/dma_iproc_pax_common.h:169]: Lack of prefix here is a good example, since this now looks like a Kconfig define (CONFIG_INTERRUPT_STATUS), which is confusing.

File: drivers/dma/dma_iproc_pax.h
Comment (olofj) [inline drivers/dma/dma_iproc_pax.h:65]: It looks like your naming is "..pax.h" for v1, "...pax_v2.h" for v2, and "...pax_common.h" for shared pieces -- it's slightly confusing compared to using "...pax_v1.h" and then "...pax.h" for shared. But that's just a minor comment.

The other concern is the duplication of definitions, with structs having different members when they come from different files, instead of being in one file where it is more obvious that there are two variants. But even then, it seems like the v2 variants are supersets, i.e. they carry forward the same fields but make use of some of the reserved bits. Would it make sense to use the v2 structs everywhere, just not touching the v2 fields in the v1 code paths?

File: dts/bindings/dma/brcm,iproc-pax-dma.yaml
Comment (olofj) [inline dts/bindings/dma/brcm,iproc-pax-dma.yaml:8]: Shouldn't you have different compatibles for v1/v2, given that they are programmed differently?

On that note: Is it possible to operate v2 in v1 mode, such that the v1 driver will work against that hardware? So should v2 be a more specific compatible, or are they mutually exclusive and need separate toplevel compatible?

File: drivers/dma/dma_iproc_pax_v2.c
Comment (olofj) [inline drivers/dma/dma_iproc_pax_v2.c:773]: This is probably better in a separate function that's always called, but with an implemention that is an empty static inline if the config option is not set. That way the code flow here doesn't need an ifdef.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (olofj) [inline drivers/dma/dma_iproc_pax_v2.c:788]: Same here, if you have a "wait_for_completion" helper, that contains either code path.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (olofj) [inline drivers/dma/dma_iproc_pax_v2.c:557]: Doing this as an if/else, but doing completely different things in the two blocks, can be confusing. It's easier to follow the code if you do two independent ifdefs.

File: drivers/dma/dma_iproc_pax_common.h
Comment (olofj) [inline drivers/dma/dma_iproc_pax_common.h:132]: Nit: Slightly odd to have a MixedCaps define in the middle of the rest.

**Review by olofj — CHANGES_REQUESTED** [review]

File: drivers/dma/dma_iproc_pax.h
Comment (xxie24) [inline drivers/dma/dma_iproc_pax.h:31]: what is the calculation behind the 16MBytes limits?

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [inline drivers/dma/dma_iproc_pax_v2.c:317]: why 4 rings here?

File: drivers/dma/CMakeLists.txt
Comment (raveenp) [reply drivers/dma/CMakeLists.txt:16]: Hi, 
There's a correction here, it should be

+zephyr_library_sources_ifdef(CONFIG_DMA_IPROC_PAX	dma_iproc_pax.c)
+zephyr_library_sources_ifdef(CONFIG_DMA_IPROC_PAX_V2	dma_iproc_pax_v2.c)

V1 and V2 are hardware needs to be programmed differently so the driver's are different.

So I feel keeping different CONFIG option is necessary here.

File: drivers/dma/Kconfig.iproc_pax
Comment (raveenp) [reply drivers/dma/Kconfig.iproc_pax:14]: Didn't fully understand your comment. As explained above DMA_IPROC_PAX_V2 is defined to address V2 hardware and DMA_IPROC_PAX is defined to address V1. So having 2 CMake entry is needed to differentiate.

File: drivers/dma/dma_iproc_pax_common.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_common.h:15]: Sure, let me fix. Thanks

File: drivers/dma/dma_iproc_pax_common.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_common.h:169]: I understand, let me fix it new patchset.

File: drivers/dma/dma_iproc_pax.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax.h:31]: Hi,
In V1 RM version, length field is of 25 bits in DMA Header descriptor.
So hardware can support upto 32MB of data length per BD, but we have set it to 16MB.

In case of V2 RM version, length field  in src/dst descriptor is modified to 15 bits.
So with this it can support max length of 32KB per BD.

But we do have an option to use Mega SRC and Mega DST which will support 16*LENGTH bits, so with this max 16*32KB = 512KB.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:317]: V1 RM supports 4 rings and V2 supports 8 rings.

We can set it to 8 but only 4 is used as of now.

File: drivers/dma/dma_iproc_pax.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax.h:65]: Yes, having "..pax_v1.h" and "..pax_v2.h" is better for clear understanding.

V1 and V2 RM hardware versions have different BD formats, so kept the different definitions.
V1 and V2 needs slightly different programming sequence as well, so the idea of having different driver files.

File: dts/bindings/dma/brcm,iproc-pax-dma.yaml
Comment (raveenp) [reply dts/bindings/dma/brcm,iproc-pax-dma.yaml:8]: Different compatible is needed for  V1 and V2. Yes, they are programmed differently.

Both are mutually exclusive, not possible to operate V2 in V1 mode.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:773]: Sure, wil fix it.

File: drivers/dma/dma_iproc_pax_common.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_common.h:132]: ack, will fix it.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:557]: ack, will fix it.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:788]: ack, will fix it.

File: drivers/dma/CMakeLists.txt
Comment (raveenp) [reply drivers/dma/CMakeLists.txt]: Please look at updated patch.

File: drivers/dma/dma_iproc_pax_common.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_common.h]: Used prefixes based on register category and where necessary PAX_DMA*

File: dts/bindings/dma/brcm,iproc-pax-dma.yaml
Comment (raveenp) [reply dts/bindings/dma/brcm,iproc-pax-dma.yaml]: Added two different compatible strings though dts entries perspective there's no distinct change required but driver point of view it's 2 different version of drivers.

**Review by Abhishek-brcm — APPROVED** [review]
LGTM

**Comment by raveenp** [comment]
@olofj @xxie24 I addressed your comments where I could do and provided my feedback on requested changes, could you please review again ?

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [inline drivers/dma/dma_iproc_pax_v2.c:804]: 1) it seems like the code are 90% duplicated in mega_len and non_mega_len case. The main difference is "is_mega" and size/offset calculation. Can we refactor this part inito one 1 function?

2) I am curious: each header can have 15 pair of src and dest. Why we need create extra packets here?

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [inline drivers/dma/dma_iproc_pax_v2.c:331]: for doorbell mode to work, don't we need enable the AE locking, and set the group ring count to 1?

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [reply drivers/dma/dma_iproc_pax_v2.c:345]: can you provide 8 ring settings as well?

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [inline drivers/dma/dma_iproc_pax_v2.c:216]: Just for information purpose, I am using the door bell mode, but I am also setting toggle bit in my driver, is there is negative effects?

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:804]: 1.  Yes, Sure. Let me refactor the code here.

2.  I understand, with our experience on Valkyrie extra header packet didn't impact much on the performance.
     Will try to assess performance difference on Viper and enhance the driver in a separate PR.
 
Incase of Non Mega src/dst descriptors,
Header BD's =  SG list length/15.
Non header BD's = SG list length % 15
So its easy to generate BD's here.

In case of Mega SRC/DST descriptors, Above calculation will not work as each sglist can come with any size and need to be split into different Mega SRC/DST descriptors. So descriptor generation is not straight forward and any additional code to handle will add burden on the CPU.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:331]: AE locking is needed if packet extension needs to be supported otherwise for doorbell mode its not required.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:216]: There's no side effect, RM will ignore the toggle bit.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:846]: Hi Xin Xie,

I have now added support to handle dma_block_config's list or scatter/gather list handling in dma_iproc_pax_v2.c driver.
And also handling 15 pair of PCIE/AXI descriptors in one header BD. Please review the same.

**Comment by raveenp** [comment]
Hi Xin Xie,
I have now added support to handle dma_block_config's list or scatter/gather list handling in dma_iproc_pax_v2.c driver,
And also handling 15 pair of PCIE/AXI descriptors in one header BD. Please review the same.

File: drivers/dma/dma_iproc_pax.h
Comment (olofj) [inline drivers/dma/dma_iproc_pax.h:387]: As mentioned in the use location of this, more clarity is needed here -- what's the intent of this macro, and what are the semantics? (even if it's not implemented right now).

File: drivers/dma/dma_iproc_pax.h
Comment (olofj) [inline drivers/dma/dma_iproc_pax.h:421]: I know I asked for prefixing types/defines to not pollute global namespace, "dma_iproc_pax" is a bit on the verbose side and "pax" would probably suffice, but I'm not going to ask you to change it.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (olofj) [inline drivers/dma/dma_iproc_pax_v2.c:450]: There's a missing unlock here.

File: dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml
Comment (olofj) [inline dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml:33]: This block only provides one interrupt, right? Might make sense to clarify that here.

File: dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml
Comment (olofj) [inline dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml:19]: This isn't memory that is part of this peripheral, right? Instead, it describes preallocated memory on the system?

**Review by olofj — COMMENTED** [review]
These are mostly nits, with a few real bugs around unlocking on failure paths. I suggest using `ret` variable and `goto out` patterns to avoid this.

File: drivers/dma/dma_iproc_pax_v1.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v1.c:62]: Sure. Thanks.  Will do.

File: drivers/dma/Kconfig.iproc_pax
Comment (raveenp) [reply drivers/dma/Kconfig.iproc_pax:8]: will fix it in the next patch.

File: drivers/dma/dma_iproc_pax_v1.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v1.c:132]: ok. will add comments.

File: dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml
Comment (raveenp) [reply dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml:19]: This memory is not part of peripheral, it's pre allocated memory on the system.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:450]: ack.

File: drivers/dma/dma_iproc_pax_v1.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v1.c:375]: ack, will fix it.

File: dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml
Comment (raveenp) [reply dts/bindings/dma/brcm,iproc-pax-dma-v2.yaml:33]: ack, will fix.

File: drivers/dma/dma_iproc_pax_v1.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v1.c:205]: dma_mb() is expected to replace "dsb" or data memory barrier kind of instruction to ensure all the writes to memory have reached, Before hardware access the BD memory.

File: drivers/dma/dma_iproc_pax.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax.h:421]: Followed naming convention lookin at other dma driver's, I understand its bit verbose.  For register definition macros, as explained earlier register name + bit field name together is becoming too lengthy if I add some common prefix.

But as such register macros are classified according to the sub blocks within Ring manager like: RING, RM_COMM, RM_AEx. Other driver specific macros are named with prefix PAX_DMA.

**Comment by raveenp** [comment]
@olofj @xxie24 :  Addressed your review comments, please have a look.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [inline drivers/dma/dma_iproc_pax_v2.c:680]: If the input transfer is divided into multiple packets, the 1st packet will trigger an ISR upon finish. The code here will unblock the start_xfer function with partial data xfer complete. 

If we use extended packet, we can avoid this issue. Let me know if you see a better option.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:787]: 1. non_hdr_bd_count seems to me a local information, we could set it as local variable in dma_iproc_pax_gen_descs, and pass it as a pointer to this function.
[raveenp] 
Current driver design is such that, Hardware ring is disabled until all the packets[total_pkts]/descriptors are generated in the memory
corresponding to the dma_block_config list/sg list passed as part of one dma_config API call.

No intermediate ISR is kicked in for the partial packets, as RM will be programmed to generated
interrupt after processing "total_pkts".

So I had kept in the ring structure to clear at the ISR.

But as per your suggestion it makes more sense to keep it as local variable. Will make the change.

2. non_hdr_bd_count is get cleared by ISR, so we also need some sort spin_lock to protect it which is headache.
[raveenp] Yes. Better to keep it as local variable.

3. In theory, we should building multiple packets on the ring before any of them finish. So putting non_hrd_bd_count into the ring will limit this feature.
[raveenp] Yes, keeping it as local variable makes it clear. will change the code.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:816]: Ack, will fix it in the next patch set.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:680]: Ring can be made inactive before generating all the packets corresponding to input transfer, Once BD's are populated in the memory- program the "total_pkt_count/header bd's count" count into RING_CMPL_WR_PTR_DDR_CONTROL so that RM generates interrupt only after processing "total_pkt_count/header bds' count".

Comment (xxie24) [reply drivers/dma/dma_iproc_pax_v2.c:680]: Right. Use set_pkt_count() will also avoid multiple interrupts for 1 xfer, so you don't need use extended packet.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [reply drivers/dma/dma_iproc_pax_v2.c:787]: A suggestion: if you put non_hdr_bd_count as local stack variable, you can also consider return "total_pkt_count"  as return value to the caller function, and pass it to set_pkt_count(). This way you don't need a global variable of total_pkt_count.

I have not thought through all the details of this, but seems doable.

Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:787]: In config() API call, descriptors are populated  and overall "total_pkt_count" is derived.
In start () API set_pkt_count() is called, ring is enabled to process the packets, so total_pkt_count is needed to be part of ring structure also In case of poll mode, start() needs to know how many packets to poll for completion.

File: drivers/dma/dma_iproc_pax.h
Comment (raveenp) [reply drivers/dma/dma_iproc_pax.h:387]: here dma_mb() should implement a macro to enable data memory barrier, to ensure writes to memory happen.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (xxie24) [inline drivers/dma/dma_iproc_pax_v2.c:783]: Why it is pointer to pointer?

**Review by xxie24 — APPROVED** [review]
please provide comment on why  `**non_hdr_bd_count` is used in dma_iproc_pax_gen_desc, and see if it is possible to just use  `*non_hdr_bd_count` instead. 

Otherwise, I see no other issue.

File: drivers/dma/dma_iproc_pax_v2.c
Comment (raveenp) [reply drivers/dma/dma_iproc_pax_v2.c:783]: It should work with single pointer, will change.

**Comment by raveenp** [comment]
@xxie24 Addressed comment related to **non_hdr_bd_count, made single pointer.

**Comment by raveenp** [comment]
@olofj Could you please review and provide feedback.

**Review by sandeepbrcm — APPROVED** [review]
LGTM

**Comment by olofj** [comment]
Hi, my outstanding comments seem to have been addressed, and I'm fine with this driver going in. I believe it still needs review from a DMA subsystem maintainer before merging though.

**Review by olofj — APPROVED** [review]
Hi, per above comment (used comment instead of review interface, github is slightly confusing on that). I'm OK with the state of this driver for merge at this point, my previous comments have been addressed to my satisfaction.

**Pull request merged**
